Add option to show all deck stats by default

smasher816 opened the pull request:
## Pull Request template

## Purpose / Description
This option is useful to anyone who wants to know "how many cards did I
study today (in total), and how many cards are due tomorrow (in total)".
In this situation it can save a few clicks and scrolls which can be slow
on e-ink displays.

## Fixes
If using a single "To Review" filtered deck then stats are pretty
meaningless. For instance "Studied 3 cards in 0 minutes today". To
find the real statistics after studying with a filtered deck I have
to select "All decks" and then I correctly see "Studied 491 cards in
25 minutes today".

## Approach
This option lets you bypass the empty filtered deck stats and go directly to the stats you are interested in (this new behavior is opt-in)

## How Has This Been Tested?

Deploy to android emulator:
 - Review a deck, open nav drawer, click stats, see last deck stats
 - Enable new "Show statistics for all decks by default" advanced setting
 - Review a deck, open nav drawer, click stats, see all decks stats

## Learning (optional, can help others)
Search strings for existing setting name

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

smasher816 commented:
Some more background context on why I think this is needed:
I believe that cards get moved into and out of filtered decks (when rebuilt, and then when reviewed). Because of this, after reviewing a filtered the deck, it is left empty until you rebuild it again the next day. If the deck is then currently empty, then that means that there are no cards to look up statistics for.
I'm not sure if there is a way to get historical information about what cards used to be in a filtered deck, but I suspect this isn't tracked as I also have to do the same "switch to all decks" juggle on the ipad app and desktop app.

mikehardy reviewed (approved):
I actually like this because I study a few main decks daily (or semi-daily, I'm behind...) and I *always* want the stats for all study all the time as a personal preference

Open to disagreement, if any, from other maintainers but I like this and would merge

BrayanDSO commented:
It looks like a nice idea.
Maybe I'm being picky, but I think the new strings shouldn't be below the `<!-- Advanced statistics settings -->` comment, because advanced statistics is a feature by itself and this may create some confusion

mikehardy commented:
I think that's a good point @BrayanDSO

smasher816 commented:
Fixed Brayan's suggestion

Arthur-Milchior commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -197,6 +197,9 @@
     <string name="media_import_allow_all_files" maxLength="41">Allow all files in media imports</string>
     <string name="media_import_allow_all_files_summ">If Android doesn’t recognize a media file</string>
+    <!-- Statistics settings -->
+    <string name="statistics_all_decks_title" maxLength="41">Show statistics for all decks by default</string>
```
I don't think it's clear what "default" mean. This option makes sens only if you know what is the current behavior.

I believe in this case it would be better to have a setting with two options.
> When you open statistics you see
> * All decks
> * Current deck

And let the user select between those two named choices

mikehardy replied:
@Arthur-Milchior I think the next string is where the extra detail goes where the user learns what default means, with explanation of the opposite, quoted here: `If disabled defaults to the last selected deck` - with those two strings presented in the user interface right next to each other, this seems understandable to me, I'm not sure how I'd change it to make it more clear. It would require a fair chunk of preferences code in order to expand a boolean / checkbox pref to a radio button with two options that represents a boolean in the actual prefs file. Not sure it's worth it

Arthur-Milchior replied:
Then I'd love to see the screenshot, because it was not clear to me how the second text is.
Also

david-allison replied:
@mikehardy Just to note: a `ListPreference` wouldn't be much code/work. I agree that the UX would be better.

mikehardy replied:
Ahh - my Preference skills are obviously lacking, didn't know there would be something out of the box and thought we'd have to write some custom pref listener code. That's two people that want the UX different, I'll defer then - sounds like a ListPreference with the two values? and a screenshot for translators...

Arthur-Milchior replied:
I beg your pardon @mikehardy, I guess that, as David did, I should have checked the name of the feature. I thought it was clear I was only asking for a classic UI interface for multiple choice. Not anything new.

Arthur-Milchior reviewed (commented):
I confirm that there is nothing that allow to see in which deck a card was when reviewed. I just double checked the revlog table.

Also, I'd love to get screenshots from the settings because we must upload it for translators.

david-allison reviewed (approved):
LGTM

mikehardy commented:
Current status (unburying the comment from a deeper thread) https://github.com/ankidroid/Anki-Android/pull/10421#discussion_r818998359

> That's two people that want the UX different, I'll defer then - sounds like a ListPreference with the two values? and a screenshot for translators...

smasher816 commented:
I have swapped to a ListPreference with case statement as suggested.

mikehardy commented:
@smasher816 thank  you for coming back around on this one! Much appreciated
The CI failure on emulator tests had nothing to do with this PR, the emulator in GitHub's data center just didn't start up correctly. I restarted it.

However there does seem to be some legitimate issue with the lint tests vs the strings?

```

/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/07-cardbrowser.xml:28: Error: Duplicate string value "All decks", used in card_browser_all_decks, stats_default_deck_all [DuplicateCrowdInStrings from com.ichi2.anki.lint]
    <string name="card_browser_all_decks">All decks</string>
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:203: Duplicates value in card_browser_all_decks. Add a comment attribute on both strings to explain this duplication
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:201: Error: Preference title 'stats_default_deck_title' is missing "maxLength=41" attribute [PreferencesTitleMaxLengthAttr from com.ichi2.anki.lint]
    <string name="stats_default_deck_title">Default deck when opening statistics</string>
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors, 0 warnings
```

mikehardy commented on AnkiDroid/src/main/res/xml/preferences_advanced.xml:
```diff
@@ -155,6 +155,12 @@
                 android:key="double_scrolling"
                 android:summary="@string/double_scrolling_gap_summ"
                 android:title="@string/double_scrolling_gap" />
+            <ListPreference
```
that looks pretty easy, not sure I was afraid it would be hard and require some custom prefs code, this is nice and clean

Arthur-Milchior replied:
I fear you forgot a "why" after "not sure"

smasher816 commented:
> However there does seem to be some legitimate issue with the lint tests vs the strings?
> 
> ```
> 
> /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/07-cardbrowser.xml:28: Error: Duplicate string value "All decks", used in card_browser_all_decks, stats_default_deck_all [DuplicateCrowdInStrings from com.ichi2.anki.lint]
>     <string name="card_browser_all_decks">All decks</string>
>     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:203: Duplicates value in card_browser_all_decks. Add a comment attribute on both strings to explain this duplication
> /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:201: Error: Preference title 'stats_default_deck_title' is missing "maxLength=41" attribute [PreferencesTitleMaxLengthAttr from com.ichi2.anki.lint]
>     <string name="stats_default_deck_title">Default deck when opening statistics</string>
>     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 2 errors, 0 warnings
> ```

Not sure how to run the tests, but I think I fixed things.
1) Added the max length value (was currently at 36 chars, so it should be a non-issue)
2) Adjusted the text so it is distinct. These should likely stay separate strings and the fact that it collided with `card_browser_all_decks` was merely a coincidence

mikehardy commented:
I pulled this locally and ran it, checked the default setting / behavior (same as previous behavior :heavy_check_mark: ) and then switched it to all decks (my preference personally too!) and it was beautiful.

mikehardy reviewed (approved):
A thing of beauty

mikehardy commented:
merging despite e2e failure - it's the "semi-normal" flake where the emulator doesn't boot in the CI environment sometimes. I ran the tests locally though and they all passed.

smasher816 commented:
Will this also get a signed pre-release to install on devices?

mikehardy commented:
It definitely will, I run the release for a new alpha at a cadence of about one every 3-4 days and I did one about 24hrs ago, so probably Saturday or Sunday?

mikehardy commented:
Not quite clockwork but reasonably close ;-) https://github.com/ankidroid/Anki-Android/actions/workflows/publish.yml

mikehardy commented:
Hi there! MARCH OpenCollective Notice

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for [March 2022 submissions](https://github.com/ankidroid/Anki-Android/pulls?q=is%3Apr+merged%3A2022-03-01..2022-03-31)

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for in the month of March.

Please note that GSoC contributions are okay for this process. Our philosophy is that our users have donated to AnkiDroid for *all* contributions. The only PRs that will not go through the OpenCollective process are ones *directly* related to a GSoC project from a selected participant, since those receive a stipend from GSoC itself.

Please understand that this means the monthly budget we have for payments might be stretched a bit and the cap on payments-per-person may be lower, but we try to make our process as fair and transparent as possible, we just need your understanding.

Thanks!
